### PR TITLE
Feature/mip 309

### PR DESCRIPTION
**Pull request by thedevnisarg**
After code cleanup and refactor.

### CODE REVIEW

File: src/main/groovy/com/avioconsulting/mule/MuleLinterCli.groovy [hunk 1]
```diff
@@ -1,27 +1,25 @@
...
 
     @Override
     void run() {
-        File appDir = new File(appDir)
-        Application app = new Application(appDir)
-        System.out.println(app.hasGitIgnore())
-        System.out.println(app.hasFile('test/file.txt'))
+        // Parse application to build out structure
+        MuleLinter ml = new MuleLinter(appDir)
+        ml.runLinter()
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/MuleLinterCli.groovy:22]: Need to handle invalid app directory/not found more cleanly.  Currently an exception is just thrown:

adesjard@ac0050:~/code/mulesoft/avio/mule-linter/mule-linter/build/libs$ java -jar mule-linter-1.0-SNAPSHOT-all.jar -d /sdgfsfdgfds -r asdfsa
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.ReflectionUtils (file:/home/adesjard/code/mulesoft/avio/mule-linter/mule-linter/build/libs/mule-linter-1.0-SNAPSHOT-all.jar) to constructor java.io.FileNotFoundException(java.lang.String,java.lang.String)
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.ReflectionUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
java.io.FileNotFoundException: /sdgfsfdgfds (Application directory does not exists.)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:72)
        at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:237)

File: src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy [hunk 2]
```diff
@@ -2,23 +2,34 @@ package com.avioconsulting.mule.linter.model
...
+    File getApplicationPath() {
+        return applicationPath
+    }
 
-        return true
+    void setApplicationPath(File applicationPath) {
+        this.applicationPath = applicationPath
+    }
+
+    PomFile getPomFile() {
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy:26]: We said this should be created as a private variable in the constructor

File: src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy [hunk 3]
```diff
@@ -2,23 +2,34 @@ package com.avioconsulting.mule.linter.model
...
+        return new PomFile(applicationPath, POM_FILE)
     }
 
-    Boolean hasGitIgnore() {
-        return hasFile('.gitignore')
+    Boolean hasFile(String filename) {
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy:30]: We might not want to check for file length if we're just checking a file exists.  That might be another method on top that also checks if there's contents

File: src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy [hunk 4]
```diff
@@ -2,23 +2,34 @@ package com.avioconsulting.mule.linter.model
...
+    static final String APPLICATION_DOES_NOT_EXIST = 'Application directory does not exists.'
+    static final String POM_FILE = 'pom.xml'
+
     File applicationPath
+    List<ProjectFile> files = []
 
     Application(File applicationPath) {
         this.applicationPath = applicationPath
+        if (!this.applicationPath.exists()) {
+            throw new FileNotFoundException(applicationPath.absolutePath, APPLICATION_DOES_NOT_EXIST)
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy:14]: The 2 argument constructor is private, should be using the 1 argument public constructor

File: src/main/groovy/com/avioconsulting/mule/linter/model/ProjectFile.groovy [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package com.avioconsulting.mule.linter.model
+
+class ProjectFile {
+
+    String name
+    String type
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/ProjectFile.groovy:6]: These are never set by the PomFile, what are they for?

File: src/main/groovy/com/avioconsulting/mule/linter/model/PomFile.groovy [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    PomFile(File application, String fileName) {
+        this(new File(application, fileName))
+    }
+
+    Boolean doesExist() {
+        return exists
+    }
+
+    String getProperty(String propertyName) {
+        return exists ? model?.properties?.getProperty(propertyName) : ''
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/PomFile.groovy:31]: Should differentiate between pom not existing, property not found, and property having an empty string value which is valid

File: src/main/groovy/com/avioconsulting/mule/linter/model/PomFile.groovy [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    private static Model parseModel(File pomFile) {
+        Model model = null
+        FileReader reader
+        MavenXpp3Reader mavenReader = new MavenXpp3Reader()
+        try {
+            reader = new FileReader(pomFile)
+            model = mavenReader.read(reader)
+            model.pomFile = pomFile
+        } catch (IOException | XmlPullParserException ex) {
+            ex.printStackTrace()
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/PomFile.groovy:43]: Should be using a Logger, not just printing stacktraces.  Also do we not want to return an exception if we can't parse the model?

File: src/main/groovy/com/avioconsulting/mule/linter/model/RuleExecutor.groovy [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        Integer count = 0
+        results.each { violation ->
+            outputStream.write("    $violation.rule.severity: $violation.fileName ".bytes)
+            outputStream.write((violation.lineNumber > 0 ? "( $violation.lineNumber ) " : '').bytes)
+            outputStream.write("$violation.message \n".bytes)
+        }
+        outputStream.write("\nFound a total of $count violations.\n".bytes)
+        outputStream.flush()
+        outputStream.close()
+    }
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/model/RuleExecutor.groovy:32]: Consider using withPrintWriter which is a Groovy enhancement to OutputStream:  http://docs.groovy-lang.org/latest/html/groovy-jdk/java/io/OutputStream.html

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/PomExistsRule.groovy [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        this.ruleName = RULE_NAME
+    }
+
+    @Override
+    List<RuleViolation> execute(Application app) {
+        List<RuleViolation> violations = []
+
+        PomFile pomFile = app.pomFile
+        if (pomFile == null || !pomFile.doesExist()) {
+            violations.add(new RuleViolation(this, app.applicationPath.toString(), 0, FILE_NOT_EXISTS))
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/rule/pom/PomExistsRule.groovy:25]: Shouldn't the location be the pomFile absolute path, not just the app directory?

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/PomPropertyValueRule.groovy [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    List<RuleViolation> execute(Application app) {
+        List<RuleViolation> violations = []
+
+        String pomPropertyValue = app.pomFile.getProperty(propertyName)
+        if (!pomPropertyValue.equalsIgnoreCase(propertyValue)) {
+            violations.add(new RuleViolation(this, app.pomFile.getFile().absolutePath,
+                    0, propertyName + RULE_VIOLATION_MESSAGE))
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/rule/pom/PomPropertyValueRule.groovy:37]: Need actual line from pom.xml where the property exists for the violation, unless it doesn't exist.  We should probably default to where the properties block is in that case

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MunitVersionRule.groovy [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import com.avioconsulting.mule.linter.model.Application
+import com.avioconsulting.mule.linter.model.RuleViolation
+
+class MunitVersionRule extends PomPropertyValueRule {
+
+    static final String RULE_ID = 'MUNIT_VERSION'
+    static final String PROPERTY_NAME = 'munit.version'
+    static final String RULE_NAME = 'munit.version maven property match'
+
+    String version
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MunitVersionRule.groovy:12]: Why are we saving the version when it isn't used for anything after its passed to the super() constructor?

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleRuntimeVersionRule.groovy [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import com.avioconsulting.mule.linter.model.Application
+import com.avioconsulting.mule.linter.model.RuleViolation
+
+class MuleRuntimeVersionRule extends PomPropertyValueRule {
+
+    static final String RULE_ID = 'MULE_RUNTIME'
+    static final String PROPERTY_NAME = 'app.runtime'
+    static final String RULE_NAME = 'app.runtime maven property match'
+
+    String version
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleRuntimeVersionRule.groovy:12]: Why are we saving the version when it isn't used for anything after its passed to the super() constructor?

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleMavenPluginVersionRule.groovy [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import com.avioconsulting.mule.linter.model.Application
+import com.avioconsulting.mule.linter.model.RuleViolation
+
+class MuleMavenPluginVersionRule extends PomPropertyValueRule {
+
+    static final String RULE_ID = 'MULE_MAVEN_PLUGIN'
+    static final String PROPERTY_NAME = 'mule.maven.plugin.version'
+    static final String RULE_NAME = 'mule.maven.plugin.version maven property match'
+
+    String version
```
Comment (adesjardin) [inline src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleMavenPluginVersionRule.groovy:12]: Why are we saving the version when it isn't used for anything after its passed to the super() constructor?

Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleMavenPluginVersionRule.groovy:12]: Right not needed. Removing

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleRuntimeVersionRule.groovy (hunk 12 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MuleRuntimeVersionRule.groovy:12]: Right not needed. Removing

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MunitVersionRule.groovy (hunk 11 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/rule/pom/MunitVersionRule.groovy:12]: Right not needed. Removing

File: src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy (hunk 3 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy:30]: checking of file length removed.

File: src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy (hunk 4 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy:14]: changed to use 1 arg constructor.

File: src/main/groovy/com/avioconsulting/mule/linter/model/ProjectFile.groovy (hunk 5 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/model/ProjectFile.groovy:6]: Not used yet, added for future use. Removing now we will add in future when needed.

File: src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy (hunk 2 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/model/Application.groovy:26]: added private variable and returning that now.

File: src/main/groovy/com/avioconsulting/mule/linter/rule/pom/PomPropertyValueRule.groovy (hunk 10 above)
Comment (thedevnisarg) [reply src/main/groovy/com/avioconsulting/mule/linter/rule/pom/PomPropertyValueRule.groovy:37]: Maven library is parsing the xml, so we dont know the actual line number where it should belong.

**Pull request merged**